addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -1283,11 +1283,301 @@ export class OdooEditor extends EventTarget {
+                            restoreEndSplit(lastChild, docE.firstElementChild);
+                        }
+                        // full restoration disregarding unremovables
+                        docE.prepend(...children);
+                        docE.normalize();
+                    } else {
+                        // editable
+                        if (unremovableChildren.length) {
+                            restoreEndSplitUnremovables(unremovableChildren, fragE, docE);
+                        } else {
+                            // no unremovable child
+                            const children = [];
+                            ensureEmptyPElement(children);
+                            docE.prepend(...children);
+                        }
+                    }
+                } else {
+                    // the last unremovable is not a cloned node
+                    previousUnremovables = unremovables;
+                }
+                // main algo
+                const unremovableRoots = [];
+                restoreUnremovables(previousUnremovables, unremovableRoots, docContainer.isContentEditable);
+                docContainer.prepend(...unremovableRoots);
+            };
+
+            // recover and test if the first batch of nodes is split and needs to be partially restored
+            let batch = parentedGrouping.shift();
+            if (batch && batch.unremovables.length) {
+                if (batch.parent.oid) {
+                    parentedGrouping.unshift(batch);
+                } else {
+                    batch.unremovables.reverse();
+                    restoreStartSplitUnremovables(batch.unremovables, startRoot, docStartRoot);
+                }
+            }
+            // recover and test if the last batch of nodes needs to be partially restored
+            batch = parentedGrouping.pop();
+            if (batch && batch.unremovables.length) {
+                if (batch.parent.oid) {
+                    parentedGrouping.push(batch);
+                } else {
+                    batch.unremovables.reverse();
+                    restoreEndSplitUnremovables(batch.unremovables, endRoot, docEndRoot);
+                }
+            }
+            // handle every other batch of nodes that were completely removed from document
+            const unremovableRoots = [];
+            for (batch of parentedGrouping) {
+                batch.unremovables.reverse();
+                restoreUnremovables(batch.unremovables, unremovableRoots, docStartRoot.parentElement.isContentEditable);
+            }
+
+            // restore nodes to the document
+            if (docStartRoot === this.editable) {
+                this.editable.prepend(...unremovableRoots);
+            } else if (docEndRoot === this.editable) {
+                this.editable.append(...unremovableRoots);
+            } else {
+                docEndRoot.before(...unremovableRoots);
+            }
+        }
+
```

Comment:
You did a lot of work on this as it seems, I'm quite impressed. You seem to have thoroughly covered a lot of different cases, so now that would be quite sad if someone had to make a change to fix a bug and actually broke other cases by mistake. Therefore, I suggest adding through tests for this feature.